# MKTG 2120 Homework 4

## Question 1

### 1a.

Using your logistic regression model from Assignment 2, compute the CLV for the
following customers, assuming an interest rate of zero, a monthly “maintenance cost” of
servicing the customer of $5, and assuming the probability of churn stays fixed over time
(i.e., they don’t change their contracts, and the tenure variable does not change):

  i. A customer who has been with the company for 20 months (tenure = 20), who
has PhoneService, who has “Fiber optic” InternetService, a “One year” Contract,
spends $60 per month (MonthlyCharges = 60), and has signed up for
PaperlessBilling. Note: You can also assume that MonthlyCharges is how much the
customer is spending each month, i.e., the revenue from that customer. (1 point)

  ii. The same customer, but who has not signed up for PaperlessBilling. (1 point)

  ---

In [ ]:
library(tidyverse)

The logistic regression model we used from Assignment 2 is as follows:

In [ ]:
telco <- read.csv("/content/sample_data/telco_data.csv")
fit_2 <- glm(Churn ~ tenure + PhoneService + InternetService + Contract + MonthlyCharges + PaperlessBilling, data = telco, family = "binomial")
summary(fit_2)


Call:
glm(formula = Churn ~ tenure + PhoneService + InternetService + 
    Contract + MonthlyCharges + PaperlessBilling, family = "binomial", 
    data = telco)

Coefficients:
                            Estimate Std. Error z value Pr(>|z|)    
(Intercept)                -0.711606   0.185938  -3.827 0.000130 ***
tenure                     -0.035382   0.002556 -13.843  < 2e-16 ***
PhoneServiceYes            -0.827540   0.167514  -4.940 7.81e-07 ***
InternetServiceFiber optic  0.831927   0.153777   5.410 6.30e-08 ***
InternetServiceNo          -0.222035   0.213438  -1.040 0.298210    
ContractOne year           -1.002658   0.130005  -7.712 1.23e-14 ***
ContractTwo year           -1.676153   0.199058  -8.420  < 2e-16 ***
MonthlyCharges              0.014895   0.004169   3.573 0.000354 ***
PaperlessBillingYes         0.454616   0.086517   5.255 1.48e-07 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Nul

Assuming that the interest rate = 0 and that the probability of churn stays constant over time, the CLV formula can be written as: CLV = r x m (1/Churn Rate). Because, 1 - r = Churn Rate, we can assume 1 - Churn = r. Therefore, for customer #1, we used the same coefficient estimates and churn rate we found from Assignment 2’s logistic regression model. This allowed us to create the logarithmic function below, which we subtracted from 1 to get r = 0.744. We then plugged that into the original formula. The margin is $55 because we gain $60 in revenue per month from the customer, but lose $5 in maintenance cost monthly. This was multiplied by (r1-r) to get CLV = $160.21.

In [ ]:
# retention rate person 1

retention_person_1 = 1-(1 / (1 + exp(-(-0.711606+(-.035382*20)-0.827540+0.831927-1.002658+(60*0.014895)+0.454616))))
retention_person_1

[1] 0.7444449

In [ ]:
# clv for person 1

clv_person_1 = 55*(retention_person_1/(1+0-retention_person_1))
clv_person_1

[1] 160.2178

In [ ]:
# retention rate person 2

retention_person_2 = 1-(1 / (1 + exp(-(-0.711606+(-.035382*20)-0.827540+0.831927-1.002658+(60*0.014895)+0))))
retention_person_2

[1] 0.8210999

In [ ]:
# clv for person 2

clv_person_2 = 55*(retention_person_2/(1+0-retention_person_2))
clv_person_2

[1] 252.4341

### 1b.

Based on your results from part a, how much should the company be willing to pay to
convert this type of customer to paperless billing? (You can ignore that the contract is one
year; assume the customer can terminate his or her contract any month.)

---

In [ ]:
clv_person_1 - clv_person_2

[1] -92.21628

Since person one’s CLV is higher than person two’s, the company should not be willing to pay anything to convert this type of customer to paperless billing. In fact, they should be willing to pay to switch back to non-paperless billing. There are likely reasons beyond the information provided to us in the assignment. For instance, it might be much easier to unsubscribe if one has paperless billing because of its availability online.

### 1c.

In reality, the assumption that the probability of churn stays fixed is a big assumption.
For example, the tenure value always changes: each period that the customer doesn’t
churn, tenure increases by one. Based on the results from your logistic regression, will
ignoring the tenure variable lead you to over- or under-estimate CLV? Explain.

---

The negative coefficient (-0.0353820) for tenure implies that longer tenure reduces churn risk. If we're ignoring tenure, churn risk will be higher than if we consider tenure. Thus, ignoring tenure causes retention rate to be under-estimated. This means that CLV will also be under-estimated because retention rate has a positive relationship with CLV. When r decreases in the numerator of the CLV formula, and (1 - r) increases in the denominator, there is a doubling effect that lowers CLV. This goes to show how loyalty and comfortability are driving factors of retention. People are scared of the unknown and prefer sticking with the status quo, which might prevent them from churning even in a disadvantageous position.

### 1d.

Again assuming an interest rate of zero, and a monthly maintenance cost of 5 bucks, how much should the company be willing to pay to acquire a customer with “Fiber optic” InternetService but no PhoneService, on a “Month-to-month” Contract, who does not sign up for PaperlessBilling, and spends $50 per month (MonthlyCharge = 50)? You can assume tenure = 1.

---

For this question, the margin lowered from $55 to $45 to account for customer 3’s spendings of $50. We found that the CLV for person three will be $19.62. This checks out since phone service and contracts are two of the strongest predictors of retention rate. Therefore, the lack of phone service and monthly contract should have (and did) reduced CLV.

In [ ]:
# retention rate person 3

retention_person_3 = 1-(1 / (1 + exp(-(-0.711606+(-.035382*1)-0+0.831927+0+(50*0.014895)+0))))
retention_person_3

[1] 0.3037108

In [ ]:
# clv for person 3

clv_person_3 = 45*(retention_person_3/(1+0-retention_person_3))
clv_person_3

[1] 19.62832

## Question 2

This question uses the Annalect data, available from Wharton Customer Analytics (WCA).  Annalect is the data management division of the Omnicom Group, a leading global advertising and marketing communications services company. This secure data set is available on Canvas under the Modules tab, and can be downloaded after you have signed the DUA. 

The dataset annalect_v3.csv includes exposures to email and online display advertisements from an international travel & tourism company, as well as conversions at the company’s website, for a subset of 5,651 active customers, tracked by cookies over a period of 60 days. Descriptions of the variables in this dataset are provided in the annalect_data_key_v3.pdf file. Annalect is interested in understanding the effectiveness of the various different campaigns they launched on behalf of this company, and in improving their attribution model. 

---

### 2a. 

Each row of the Annalect data is an impression, where a potential customer interacted with one of the company’s touchpoints. The variable converted captures whether a customer ended up either visiting the website or making a purchase after that impression. For each of the campaign types (variable campaign_type), compute the conversion rate, using the variable converted. 

Note that the conversion rate here is based on conversions per impression, since website visits are considered conversions in this dataset.

---

In [ ]:
library(dplyr)


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union




In [ ]:
data <- read.csv("/content/annalect_v3.csv")

# Calculate conversion rates
conversion_rates <- tapply(data$converted, data$campaign_type, function(x) mean(x == "Yes"))

# Print conversion rates by campaign type
for(campaign in names(conversion_rates)) {
  cat(campaign, ":", conversion_rates[campaign], "\n")
}

Direct : 0.2435581 
Display : 0.0008358409 
Email : 0.0004093328 
OrganicSearch : 0.1439563 
OrganicSocial : 0 
PaidSearch : 0.2128485 
PaidSocial : 0.3684211 


### 2b.

Using terminology we learned in class, explain the differences you found in part a: what patterns in the conversion rates align with your expectations? Is anything surprising?

---

* `PaidSocial` has the highest conversion rate (36.8%). This aligns with common expectations as paid social campaigns often effectively target demographics of interest, leading to higher conversion rates compared to other channels.

* `Direct` (24.4%), `PaidSearch` (21.3%), and `OrganicSearch` (14.4%) have relatively high conversion rates. This aligns with our expectation as these channels typically target specific audiences actively searching for the products or services the company provides, which is more likely to lead to high conversions. The high conversion of OrganicSearch may also benefit from customers who are already aware of the brand.

* `Email` and `Display` have relatively low conversion rates. This is not surprising as these channels may be used for broad targeting to increase brand awareness, capturing users at earlier stages of the customer journey where conversion is less likely to occur immediately.

* Surprisingly, `OrganicSocial` has a conversion rate of 0. This might indicate insufficient measurement / tracking capabilities or a lack of conversions attributed directly to organic social media interactions. This could be further investigated to understand if there are any technical issues or if organic social campaigns need optimization to increase conversion.

### 2c.

Using a last-click attribution rule, and assuming the company pays per impression, which campaign type is the most valuable in driving conversions? Explain your answer, and any assumptions you need to make. 

---

In [ ]:
most_valuable_campaign <- names(which.max(conversion_rates))
most_valuable_campaign

[1] "PaidSocial"

Using a last-click attribution rule, we assume that any conversion should be attributed 100% to the last touchpoint before the conversion. Here, we select the campaign type with the highest conversion rate to be the most valuable in driving conversions: `PaidSocial` (36.8%).

### 2d.

The company is considering no longer investing in paid search advertising. Test whether there is a statistically significant difference in conversion rate between the PaidSearch and OrganicSearch campaigns. Report which test you are using, what significance cutoff you are using, what the null and alternative hypotheses are, the results, and your conclusions.

---

In [ ]:
paid_search <- data %>% filter(campaign_type == "PaidSearch")
organic_search <- data %>% filter(campaign_type == "OrganicSearch")

# Perform two-sample proportion test
test_result <- prop.test(x = c(sum(paid_search$converted == "Yes"), sum(organic_search$converted == "Yes")),
                         n = c(nrow(paid_search), nrow(organic_search)))
test_result


	2-sample test for equality of proportions with continuity correction

data:  c(sum(paid_search$converted == "Yes"), sum(organic_search$converted == "Yes")) out of c(nrow(paid_search), nrow(organic_search))
X-squared = 65.412, df = 1, p-value = 6.075e-16
alternative hypothesis: two.sided
95 percent confidence interval:
 0.05210829 0.08567605
sample estimates:
   prop 1    prop 2 
0.2128485 0.1439563 


* To test for a statistically significant difference in conversion rate between `PaidSearch` and `OrganicSearch` campaigns, we can use a two-sample proportion test.

* Null hypothesis: There is no difference in conversion rate between `PaidSearch` and `OrganicSearch` campaigns.

* Alternative hypothesis: There is a difference in conversion rate between `PaidSearch` and `OrganicSearch` campaigns.

* Significance cutoff: 0.05

* P-value is smaller than 0.05, so we reject the null. There is a statistically significant difference in conversion rate between `PaidSearch` and `OrganicSearch`.

### 2d.

What can you conclude about the relative value of paid versus organic search? Should the company continue paying for sponsored search advertising? Explain your reasoning, or describe any additional data or studies you would need to come to a conclusion.

---

The two-sample proportion test results indicate a statistically significant difference in conversion rates between `PaidSearch` and `OrganicSearch` campaigns. `PaidSearch` has a higher conversion rate (21.3%) compared to `OrganicSearch` (14.4%).

However, this statistical analysis alone doesn't provide a definitive answer about whether the company should continue paying for sponsored search advertising. Additional considerations and data are needed:

* Cost of `PaidSearch`: even if `PaidSearch` has a higher conversion rate, if it's significantly more expensive, it might not be as profitable as `OrganicSearch`

* Attribution Modeling: use alternative attribution rules, if more accurate than the current last-click attribution rule, to re-consider the contribution of each marketing channel throughout the customer journey and see if the results suggest otherwise

* Short-term vs long-term lift and other effects: consider both the short- and long-term lift of `PaidSearch` vs `OrganicSearch` campaigns on conversion, as well as other effects on customer loyalty, etc.

* Qualitative Research: conduct qualitative research such as surveys or focus groups to understand customer preferences, attitudes, and behaviors regarding different types of advertisements

* Experimentation: conduct A/B testing and RCTs to directly compare the conversion effectiveness of `PaidSearch` vs other strategies

By combining these additional analyses with the statistical findings, the company can make a more informed decision about whether to continue investing in `PaidSearch` campaigns or explore other strategies.

## Question 3

The file annalect_v3_conv_exp.csv contains the same data as in the original file, but re-arranged. In this file, the rows are user/conversion pairs (e.g., user 10’s 3rd conversion). The variable “converted” tracks whether the user converted before the end of the tracking period.  The remaining columns are each of the campaign types that each user was exposed to prior to conversion (or, if they haven’t yet converted, then all the campaigns they have been exposed to until the end of the data). The campaign exposures are measured as binary variables, which equal one if the user was exposed to that campaign type, and zero otherwise. We will use this data to build an algorithmic multitouch attribution model. 

---

### 3a.

Use logistic regression to explain conversion as a function of the campaign types. Report and interpret your regression results: how do each of the different campaign types influence the probability of conversion? (3 points)

Important: when running regression analyses on this specific dataset, we cannot include an intercept in the model (see the bonus question below). To run regression in R without an intercept, you simply “add” a 0 in the model question, i.e.,


glm(converted ~ 0 + …your X’s…, family = binomial, data = data)

---

In [ ]:
# Load data
conv_exp <- read.csv("/content/annalect_v3_conv_exp.csv")

# Implement logistic regression
glm(converted ~ 0 + Direct + Display + Email + OrganicSearch + OrganicSocial +
  PaidSearch + PaidSocial, data = conv_exp, family = "binomial") %>%
    summary


Call:
glm(formula = converted ~ 0 + Direct + Display + Email + OrganicSearch + 
    OrganicSocial + PaidSearch + PaidSocial, family = "binomial", 
    data = conv_exp)

Coefficients:
               Estimate Std. Error z value Pr(>|z|)    
Direct          0.90997    0.03327  27.351  < 2e-16 ***
Display        -1.94585    0.05347 -36.389  < 2e-16 ***
Email          -2.39930    0.13121 -18.286  < 2e-16 ***
OrganicSearch  -0.21448    0.05122  -4.188 2.82e-05 ***
OrganicSocial -11.65737  135.14611  -0.086   0.9313    
PaidSearch      0.36094    0.04995   7.226 4.96e-13 ***
PaidSocial      1.21492    0.61419   1.978   0.0479 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 13159  on 9492  degrees of freedom
Residual deviance: 10292  on 9485  degrees of freedom
AIC: 10306

Number of Fisher Scoring iterations: 12


* `Display`, `Email`, and `OrganicSearch` have statistically significant negative coefficients, indicating that exposure to these campaign types decreases the log odds of conversion.

* `OrganicSocial` has a negative coefficient, but the effect is not statistically significant.

* `Direct`, `PaidSearch`, and `PaidSocial` have statistically significant positive coefficients, indicating that exposure to these campaign types increases the log odds of conversion.



### 3b.

Re-run the logistic regression model from part f, using data only for people who have already converted once before (i.e., in R, using: data[data$conversion_number > 1,]). Describe how the results change. Explain these differences using terminology from class. (Remember: conversion here means either visiting the website or making a purchase.) (3 points)

---

In [ ]:
# Load data
conv_exp_converted <- conv_exp[conv_exp$conversion_number > 1,]

# Implement logistic regression
glm(converted ~ 0 + Direct + Display + Email + OrganicSearch + OrganicSocial +
  PaidSearch + PaidSocial, data = conv_exp_converted, family = "binomial") %>%
    summary


Call:
glm(formula = converted ~ 0 + Direct + Display + Email + OrganicSearch + 
    OrganicSocial + PaidSearch + PaidSocial, family = "binomial", 
    data = conv_exp_converted)

Coefficients:
               Estimate Std. Error z value Pr(>|z|)    
Direct          1.50899    0.05398  27.954  < 2e-16 ***
Display        -2.00315    0.08340 -24.018  < 2e-16 ***
Email          -1.10716    0.31935  -3.467 0.000527 ***
OrganicSearch   0.41501    0.10225   4.059 4.94e-05 ***
OrganicSocial  -9.56291  196.96770  -0.049 0.961277    
PaidSearch      1.02723    0.09583  10.720  < 2e-16 ***
PaidSocial      1.76138    1.26865   1.388 0.165018    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 5324.8  on 3841  degrees of freedom
Residual deviance: 3872.7  on 3834  degrees of freedom
AIC: 3886.7

Number of Fisher Scoring iterations: 10


* The coefficient of `OrganicSearch` remains statistically significant but turns from negative in the previous case to now positive. This change occurs due to the diminished effect of “crowd-out” among previously converted customers. In the previous logistic regression analysis, we included all customers regardless of their conversion history with the brand, so clicks from the organic search results of a brand could sometimes be stolen from the sponsored search results of its competitors. However, if we focus on customers with conversion number greater than 1 (i.e., previously converted), the crowd-out effect could diminish against the backdrop of heightened brand awareness from the previous conversion.

* The coefficient of `PaidSocial` becomes statistically insignificant, which suggests for customers who have already converted, `PaidSocial` might not be a decisive factor for subsequent conversions.

* The coefficients of `Direct` and `PaidSearch` remain positive but the size of effect becomes larger. This reflects increased brand awareness after the previous exposure.

* The coefficients of `Display` and `Email` remain negative. This is similar to the previous case where `Display` and `Email` decrease the log-odds of conversion.

### 3c.

Explain how the results from parts f-g could be used to develop an attribution model. That is: how can we use these logistic regression models to assign credit to different touchpoints on the path-to-conversion?

---

*  Probabilistic attribution: first we need to convert the log-odds impact of each channel into the probability impact of conversion
*  Credit assignment: we assign credits to different channels where the channels that contribute more to conversion probability get more credits
*  Customer journey: since different channels have different effects on conversion probability as we observed in 2.g., we could create a more granular path-to-conversion model specific to different stages of the customer journey, which requires separate analysis based on the number of conversions
*   Coefficient Interpretation: The coefficients from the logistic regression models represent the log odds ratios of conversion associated with each campaign type. By exponentiating these coefficients, we obtain the odds ratios, which indicate the relative impact of each campaign type on conversion probability. These odds ratios can serve as weights for attributing credit to each touchpoint.



### 3d.

Critique the model in parts f-g as the basis for an algorithmic attribution model. What is missing from the model or data that you would need to do a better job?

---

- The model fails to account for other customer behaviour factors, such as demographics, time of day, and past purchase history, that could affect click/purchase conversion. We need to collect those data better to understand the effects of different channels on conversion. In doing so, we need to ensure that we use a random sample representative of its customer population to compare the results to different datasets.
- The model also needs to consider the sequence of touchpoint exposure, where primacy and recency effects matter, as well as potential interactions between touchpoints. Further investigation is needed to assign credit to different channels because proving causation requires controlling other factors and a timing sequence.
- To test the model’s robustness, we could look at result sensitivity across other attribution models such as last click, first touch, linear, time decay, and u-shaped, as each model has different assumptions about the add process, and they could impact the results.